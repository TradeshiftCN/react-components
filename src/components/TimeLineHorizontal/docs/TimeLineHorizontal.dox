/**
 * @title 基本功能
 * @renderModel left
 */
import { TimeLineHorizontal } from '@tradeshift/react-components';

const TimeLineHorizontalBasic = () => {
  return (
    <TimeLineHorizontal configData={
    {
    			items: [
    				{
    					title: '基本资料上传',
    					description: '将用于基本信息生成'
    				},
    				{
    					title: '授信资料上传',
    					description: '将用于融资授信'
    				},
    				{
    					title: '交易明细上传',
    					description: '将用于定制化融资'
    				}
    			],
    			currentState: '授信资料上传'
    		}
    } />
  )
};

@component TimeLineHorizontal
