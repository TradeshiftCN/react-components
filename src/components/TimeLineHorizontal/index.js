import React from 'react';
import PropTypes from 'prop-types';
import './index.less';

const getStateClass = (itemIndex, currentReadyStateIndex, lastIndex) => {
  if (
    itemIndex < currentReadyStateIndex &&
    itemIndex !== currentReadyStateIndex - 1
  ) {
    return 'ready';
  }
  if (itemIndex === currentReadyStateIndex - 1) {
    return 'left-of-current-state';
  }
  if (itemIndex === currentReadyStateIndex && itemIndex === lastIndex) {
    return 'current-end';
  }
  if (itemIndex === currentReadyStateIndex) {
    return 'current';
  }
  if (itemIndex > currentReadyStateIndex && itemIndex !== lastIndex) {
    return 'un-ready';
  }
  if (itemIndex === lastIndex) {
    return 'end';
  }
};

const TimeLineHorizontal = props => {
  const { configData } = props;
  return (
    <div className="time-line-container">
      {configData.items.map((item, index, arr) => {
        const currentReadyStateIndex = configData.items.findIndex(
          element => element.title === configData.currentState
        );
        let stateNode = null;
        if (index === currentReadyStateIndex) {
          stateNode = (
            <div>
              <div
                className={`circle ${getStateClass(
                  index,
                  currentReadyStateIndex,
                  arr.length - 1
                )}`}
              />
              <div className="out-circle" />
            </div>
          );
        } else {
          stateNode = (
            <div
              className={`circle ${getStateClass(
                index,
                currentReadyStateIndex,
                arr.length - 1
              )}`}
            />
          );
        }
        return (
          <div
            key={item.title}
            className={`time-line-node ${getStateClass(
              index,
              currentReadyStateIndex,
              arr.length - 1
            )}`}
          >
            {stateNode}
            <div>{item.title}</div>
            <div>({item.description})</div>
          </div>
        );
      })}
    </div>
  );
};

TimeLineHorizontal.propTypes = {
  // TimeLineHorizontal配置对象
  configData: PropTypes.object
};

export default TimeLineHorizontal;
