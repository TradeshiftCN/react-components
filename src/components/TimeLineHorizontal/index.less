@import '~ts-variables';

.time-line-container {
  display: flex;
  flex-direction: row;
  margin: @ts-unit;

  @node-box-width: @ts-unit * 9;
  @inner-circle-width: @ts-unit;

  .time-line-node {
    display: flex;
    flex-direction: column;
    align-items: center;
    position: relative;
    width: @node-box-width;
    &.ready {
      color: @ts-color-green;
    }
    &.left-of-current-state {
      color: @ts-color-green;
    }
    &.current-end {
      color: @ts-color-blue;
    }
    &.current {
      color: @ts-color-blue;
    }
    &.un-ready {
      color: @ts-color-gray;
    }
    &.end {
      color: @ts-color-gray;
    }

    .circle {
      width: @inner-circle-width;
      height: @inner-circle-width;
      -moz-border-radius: @inner-circle-width;
      -webkit-border-radius: @inner-circle-width;
      border-radius: @inner-circle-width;

      .circle-after {
        content: '';
        position: absolute;
        top: @inner-circle-width / 2;
        bottom: 0;
        left: @node-box-width / 2 + @inner-circle-width / 2;
        width: @node-box-width - @inner-circle-width;
        height: 2px;
      }

      &.ready {
        background-color: @ts-color-green;
        &::after {
          .circle-after;
          background-color: @ts-color-green;
        }
      }
      &.left-of-current-state {
        background-color: @ts-color-green;
        &::after {
          .circle-after;
          width: @node-box-width - @inner-circle-width - @diff-outer-inner / 2;
          background-color: @ts-color-green;
        }
      }
      &.current-end {
        background-color: @ts-color-blue;
      }
      &.current {
        background-color: @ts-color-blue;
        &::after {
          .circle-after;
          background-color: @ts-color-gray;
        }
      }
      &.un-ready {
        background-color: @ts-color-gray;
        &::after {
          .circle-after;
          background-color: @ts-color-gray;
        }
      }
      &.end {
        background-color: @ts-color-gray;
      }
    }

    @diff-outer-inner: 6px;
    @outer-circle-width: @ts-unit + @diff-outer-inner;
    .out-circle {
      width: @outer-circle-width;
      height: @outer-circle-width;
      -moz-border-radius: @outer-circle-width;
      -webkit-border-radius: @outer-circle-width;
      border-radius: @outer-circle-width;
      background-color: transparent;
      border: @ts-color-blue 1px solid;
      margin-top: -(@outer-circle-width + @inner-circle-width)/2;
      margin-left: -@diff-outer-inner / 2;
    }
  }
}
