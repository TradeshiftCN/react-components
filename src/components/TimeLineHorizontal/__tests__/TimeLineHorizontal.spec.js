import React from 'react';
import { shallow } from 'enzyme';
import TimeLineHorizontal from '../index';

describe('TimeLineHorizontal', () => {
	it('should renders correctly', () => {
		const props = {
			configData: {
				items: [
					{
						title: '已经完成的小圆圈',
						description: '完成啦啦啦'
					},
					{
						title: '又一个小圆圈',
						description: '好像也完成了呀'
					},
					{
						title: '当前小圆圈',
						description: '正在完成它'
					},
					{
						title: '没完成的小圆圈',
						description: '马上就完成它'
					},
					{
						title: '最后一个小圆圈',
						description: '终于最后啦啊'
					}
				],
				currentState: '当前小圆圈'
			}
		};
		const wrapper = shallow(<TimeLineHorizontal {...props} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should renders correctly when current is the end', () => {
		const props = {
			configData: {
				items: [
					{
						title: '已经完成的小圆圈',
						description: '完成啦啦啦'
					},
					{
						title: '又一个小圆圈',
						description: '好像也完成了呀'
					},
					{
						title: '最后一个小圆圈',
						description: '终于最后啦啊'
					}
				],
				currentState: '最后一个小圆圈'
			}
		};
		const wrapper = shallow(<TimeLineHorizontal {...props} />);
		expect(wrapper).toMatchSnapshot();
	});
});
