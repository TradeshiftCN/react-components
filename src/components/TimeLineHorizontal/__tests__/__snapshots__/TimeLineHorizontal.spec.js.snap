// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeLineHorizontal should renders correctly 1`] = `
<div
  className="time-line-container"
>
  <div
    className="time-line-node ready"
    key="基本资料上传"
  >
    <div
      className="circle ready"
    />
    <div>
      基本资料上传
    </div>
    <div>
      (
      将用于基本信息生成
      )
    </div>
  </div>
  <div
    className="time-line-node left-of-current-state"
    key="授信资料上传"
  >
    <div
      className="circle left-of-current-state"
    />
    <div>
      授信资料上传
    </div>
    <div>
      (
      将用于融资授信
      )
    </div>
  </div>
  <div
    className="time-line-node current-end"
    key="交易明细上传"
  >
    <div>
      <div
        className="circle current-end"
      />
      <div
        className="out-circle"
      />
    </div>
    <div>
      交易明细上传
    </div>
    <div>
      (
      将用于定制化融资
      )
    </div>
  </div>
</div>
`;
